### PR TITLE
Refactor plugins backward

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Add `pre_backward` and `post_backward` to the precision plugin. Just as we have for the training type plugin.
- The previous plugins can modify the `closure_loss`.
- Refactor the precision `backward` API to reduce code duplication across plugins and clean the structure.
- Properly type the optimizer arguments as `Optional` since they are not available during manual optimization.
- Docs fixes
- Always call the `on_after_backward` hook after backward regardless of `should_accumulate`
- Always call the `on_after_backward` hook after backward when using manual optimization
- Always call the `LightningModule.backward` hook. It was only done for Apex in manual optimization

Part of #7740 which validates that the hook order is correct

### Does your PR introduce any breaking changes ? If yes, please list them.

EXPERIMENTAL API:
- The `backward` hook of the precision plugin no longer takes a `should_accumulate` flag.
- The `backward` hook of the precision plugin no longer returns a value.
- The `pre_backward` and `post_backward` hooks of the training type plugin no longer take optimizer arguments.

STABLE API:
- `on_after_backward` was only called if `should_accumulate`. It is now called always after backward.
- `on_after_backward` was only called with the unscaled loss on mixed precision. This is no longer the case.

NOTES:
- The `LightningModule.backward` hook will still get the optimizers arguments, which are now passed through `*args, **kwargs`. So no changes.
- A follow-up PR (#8048) will implement `on_before_optimizer_step` which will have the unscaled loss and will be only called if not accumulating. So the users who were relying on the previous behaviour can just update to the new hook.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -128,6 +128,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added XLA Profiler ([#8014](https://github.com/PyTorchLightning/pytorch-lightning/pull/8014))
 
 
+- Added `PrecisionPlugin.{pre,post}_backward` ([#8328](https://github.com/PyTorchLightning/pytorch-lightning/pull/8328))
+
+
+- Added `on_load_checkpoint` and `on_save_checkpoint` hooks to the `PrecisionPlugin` base class ([#7831](https://github.com/PyTorchLightning/pytorch-lightning/pull/7831))
```
Comment (carmocca) [inline CHANGELOG.md:134]: This was in the wrong section

File: pytorch_lightning/accelerators/accelerator.py [hunk 2]
```diff
@@ -362,7 +358,7 @@ def setup_precision_plugin(self) -> None:
         model, optimizers, schedulers = self.precision_plugin.connect(self.model, self.optimizers, self.lr_schedulers)
         self.model = model
         self.optimizers = optimizers
-        self.schedulers = schedulers
+        self.lr_schedulers = schedulers
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:361]: We use `self.lr_scheduler` everywhere else. I assume it's a typo but unrelated to this PR

File: pytorch_lightning/accelerators/accelerator.py [hunk 3]
```diff
@@ -274,27 +274,23 @@ def validation_step_end(self, output: Optional[STEP_OUTPUT]) -> Optional[STEP_OU
...
-            self.lightning_module, closure_loss, optimizer, optimizer_idx, should_accumulate, *args, **kwargs
-        )
+        self.precision_plugin.backward(self.lightning_module, closure_loss, *args, **kwargs)
 
-        self.training_type_plugin.post_backward(closure_loss, should_accumulate, optimizer, optimizer_idx)
+        closure_loss = self.precision_plugin.post_backward(self.lightning_module, closure_loss)
+        self.training_type_plugin.post_backward(closure_loss)
 
-        return output
+        return closure_loss
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:293]: This is the overview of the changes in this PR

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1447,18 +1447,20 @@ def training_step(...):
         self._verify_is_manual_optimization('manual_backward')
 
         # backward
-        self.trainer.fit_loop.epoch_loop.batch_loop.backward(loss, optimizer=None, opt_idx=None, *args, **kwargs)
+        self.trainer.fit_loop.epoch_loop.batch_loop.backward(loss, None, None, *args, **kwargs)
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1450]: the `batch_loop.backward` method has these as positional, so we pass them like that

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 5]
```diff
@@ -75,27 +52,21 @@ def pre_optimizer_step(
...
-
-        if not pl_module.automatic_optimization:
-            self.scaler.unscale_(optimizer)
-            pl_module.trainer.call_hook("on_after_backward")
-            self.scaler.step(optimizer)
-            self.scaler.update()
-        else:
+        # TODO: Add `on_before_optimizer_step`
+        # self.scaler.unscale_(optimizer)
+        # pl_module.trainer.call_hook("on_before_optimizer_step")
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/native_amp.py:62]: This will be added in a follow-up

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 6]
```diff
@@ -75,27 +52,21 @@ def pre_optimizer_step(
...
-            # lambda_closure returning None indicates that backward has been skipped
-            if result is not None:
-                self.scaler.step(optimizer)
-                self.scaler.update()
-
+            if result is None:
+                # lambda_closure returning None indicates that backward has been skipped
+                return False
+        self.scaler.step(optimizer)
+        self.scaler.update()
```
Comment (tchaton) [inline pytorch_lightning/plugins/precision/native_amp.py:69]: Neat !

File: tests/plugins/test_amp_plugins.py [hunk 7]
```diff
@@ -72,6 +72,11 @@ def test_amp_apex_ddp(
 class GradientUnscaleBoringModel(BoringModel):
 
     def on_after_backward(self):
+        # TODO: replace with `on_before_optimizer_step` so we don't need to check accumulate and unscale manually
+        if self.trainer.fit_loop.should_accumulate():
+            return
+        opt = self.optimizers()
+        self.trainer.precision_plugin.scaler.unscale_(opt)
```
Comment (tchaton) [inline tests/plugins/test_amp_plugins.py:79]: When having `on_before_optimizer_step`. Let's make sure we don't unscaled twice.

Comment (carmocca) [reply tests/plugins/test_amp_plugins.py:79]: Yeah, will remember to update this in the followup

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 8]
```diff
@@ -51,43 +51,20 @@ def backward(
...
-        error = scaled_loss.__exit__(None, None, None)
-        if error:
-            raise Exception("apex unscale error")
-
-        # once backward has been applied, release graph
-        closure_loss = closure_loss.detach()
-        return closure_loss
+        opt = optimizer or model.trainer.optimizers
+        with amp.scale_loss(closure_loss, opt):
+            super().backward(model, closure_loss, optimizer, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/plugins/precision/apex_amp.py:67]: For the `PrecisionPlugin.post_backward`, yes as we want to use it to detach.

But for the `TrainingTypePlugin.post_backward`, no plugin is using it and could be removed.

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py:67]: Ok, let's remove the loss in another PR.

File: pytorch_lightning/accelerators/accelerator.py [hunk 9]
```diff
@@ -274,27 +274,23 @@ def validation_step_end(self, output: Optional[STEP_OUTPUT]) -> Optional[STEP_OU
...
-            should_accumulate: whether to accumulate gradients
         """
-        self.training_type_plugin.pre_backward(closure_loss, should_accumulate, optimizer, optimizer_idx)
+        self.training_type_plugin.pre_backward(closure_loss)
+        closure_loss = self.precision_plugin.pre_backward(self.lightning_module, closure_loss)
 
-        output = self.precision_plugin.backward(
-            self.lightning_module, closure_loss, optimizer, optimizer_idx, should_accumulate, *args, **kwargs
-        )
+        self.precision_plugin.backward(self.lightning_module, closure_loss, *args, **kwargs)
```
Comment (justusschock) [inline pytorch_lightning/accelerators/accelerator.py:288]: what about the other args here? as `optimizer`, `optimizer_idx` and `should_accumulate`? This is a BC right?

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 5 above)
Comment (justusschock) [reply pytorch_lightning/plugins/precision/native_amp.py:62]: Does this mean there is no unscaling performed if this PR is merged? That means manual optimization would break on master, right?

File: tests/plugins/test_amp_plugins.py (hunk 7 above)
Comment (justusschock) [reply tests/plugins/test_amp_plugins.py:79]: Is this also reflected in the docs? Otherwise this is a BC on master.

**Review by justusschock — APPROVED** [review]
LGTM.

Just added some notes on potential breaking changes on master, which I think are fine if the follow-up is done quickly :)

File: pytorch_lightning/accelerators/accelerator.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/accelerators/accelerator.py:288]: Optimizer and opt index are passed with args kwargs

Should accumulate wasnt passed to the LightningModule hook so can be removed as we no longer use it

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/native_amp.py:62]: From the `scaler.step` docs:

```python
        1.  Internally invokes ``unscale_(optimizer)`` (unless :meth:`unscale_` was explicitly called for ``optimizer``
            earlier in the iteration).  As part of the :meth:`unscale_`, gradients are checked for infs/NaNs.
```

So scaling is still performed, just not manually by us as we only did it to call the hook

File: tests/plugins/test_amp_plugins.py (hunk 7 above)
Comment (carmocca) [reply tests/plugins/test_amp_plugins.py:79]: I don't think the docs mentioned the skip on accumulation or that the loss was unscaled before.

I'll add these docs improvements when `on_before_optimizer_step` is added.

File: CHANGELOG.md [hunk 10]
```diff
@@ -386,6 +404,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `self.optimizers()` not returning a single optimizer if it had been wrapped ([#8326](https://github.com/PyTorchLightning/pytorch-lightning/pull/8326))
 
 
+- Fixed the `on_after_backward` hook not getting called when using manual optimization and no plugins ([#8328](https://github.com/PyTorchLightning/pytorch-lightning/pull/8328))
+
+
+- Fixed the `LightningModule.backward` hook only getting called with the `apex` plugin ([#8328](https://github.com/PyTorchLightning/pytorch-lightning/pull/8328))
```
Comment (awaelchli) [inline CHANGELOG.md:410]: I don't understand, where else would the backward happen?
Can you point me to the changes?

File: pytorch_lightning/core/lightning.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1450]: but the previous way was much clearer to read and Python is totally fine if you name the positional args.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 11]
```diff
@@ -587,26 +571,26 @@ def _check_finite(self, loss: Tensor) -> None:
...
+        **kwargs: Any,
     ) -> None:
         """Performs the backward step.
 
         Args:
             result: The output of the trainstep (including the loss value)
-            optimizer: The optimizer optimizing the gradients to call backward for
-            opt_idx: the index of the current optimizer
+            optimizer: Current optimizer being used. ``None`` if using manual optimization.
+            opt_idx: Index of the current optimizer being used. ``None`` if using manual optimization.
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:585]: yep, nice!
sorry this wasn't made clearer before. definitely caused some confusion

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 12]
```diff
@@ -51,43 +51,20 @@ def backward(
...
-            opt_idx: the optimizer index
-            should_accumulate: whether to accumulate gradients or not
-
+            optimizer: the optimizer to perform the step later on
         """
-        opt = model.trainer.optimizers if optimizer is None else optimizer
-        scaled_loss: ContextManager[Tensor] = amp.scale_loss(closure_loss, opt)
-
-        # enter apex context
-        closure_loss = scaled_loss.__enter__()
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/apex_amp.py:74]: Pretty sure this is still needed.
William experienced problems here where the `with amp.scale_loss(closure_loss, opt):` would not work and has to call manually the enter and exit. Probably safer to keep it, not to break anything!

File: CHANGELOG.md (hunk 10 above)
Comment (carmocca) [reply CHANGELOG.md:410]: Missed this:

```suggestion
- Fixed the `LightningModule.backward` hook only getting called with the `apex` plugin when using manual optimization ([#8328](https://github.com/PyTorchLightning/pytorch-lightning/pull/8328))
```

Apex was the only one calling the `LightningModule` hook:

https://github.com/PyTorchLightning/pytorch-lightning/blob/e9d0fe867f7aec14689d7069b3fa79ef047448e9/pytorch_lightning/plugins/precision/apex_amp.py#L78-L79

As it is the only plugin to override `backward` from the base class.

The base class does:

https://github.com/PyTorchLightning/pytorch-lightning/blob/e9d0fe867f7aec14689d7069b3fa79ef047448e9/pytorch_lightning/plugins/precision/precision_plugin.py#L76-L80

Which as you can see, only calls it if `automatic_optimization`

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py (hunk 9 above)
Comment (yifuwang) [reply pytorch_lightning/accelerators/accelerator.py:288]: Hi @carmocca, I noticed that in some cases the optimizer is being directly plumbed down to `loss.backward(*args, **kwargs)` through `args`, which is breaking some of our tests (bisection pointed to this PR). It'd take some time to come up with a repro since the setup is a bit complicated. Asking first in case you know what could have gone wrong.

cc: @ananthsub, @kandluis

Comment (carmocca) [reply pytorch_lightning/accelerators/accelerator.py:288]: Perhaps you are using `manual_backward(loss, optimizer)`?

The optimizer argument was removed (in 1.4,  after the deprecation process) and if you are still passing it will be treated as part of the `*args` here

Comment (yifuwang) [reply pytorch_lightning/accelerators/accelerator.py:288]: @carmocca thanks for the prompt response! Is not passing optimizer to manaul_backward the recommended fix here?

Comment (carmocca) [reply pytorch_lightning/accelerators/accelerator.py:288]: Correct. The argument was removed in #8287 and was unused since 1.2

**Comment by kandluis** [comment]
So one interesting non-BC change here is that `on_after_backward` is now called before the gradients are clipped/normalized (if `gradient_clip_val` is set in the `Trainer`). 

I want to confirm this is an intentional design choice here? For a short-term mitigation, we've changed some of our downstream modules to use the `on_before_optimizer_step` hook rather than the `on_after_backward`. 

Is this life-cycle detail documented somewhere I might have missed.

FYI @ananthsub

**Comment by carmocca** [comment]
Hi @kandluis !

You are correct, what you pointed out is an intentional design choice. The change does not look as clear by just looking at this PR since it was done in 2 separate PRs.

This is mentioned in https://github.com/PyTorchLightning/pytorch-lightning/pull/8048. See the associated CHANGELOG updates
